File: numbers/src/main/java/org/dicio/numbers/unit/Duration.java
package org.dicio.numbers.unit;

import static org.dicio.numbers.util.Utils.remainderFromRoundingToLong;
import static org.dicio.numbers.util.Utils.roundToLong;

import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;

/**
 * TODO add documentation and tests
 */
public class Duration {

    private static final long SECONDS_IN_DAY = ChronoUnit.DAYS.getDuration().getSeconds();
    private static final double DAYS_IN_MONTH
            = (double) ChronoUnit.MONTHS.getDuration().getSeconds()
            / (double) ChronoUnit.DAYS.getDuration().getSeconds(); // ~30.436875
    
    private final long nanos; // basic well-defined unit of time, fits precisely up to ~10^3 hours
    private final long days; // not stored as nanos because of e.g. summer/winter time changes
    private final long months; // not stored as days because months have different numbers of days
    private final long years; // not stored as months because of e.g. historical changes


    public Duration(final long nanos, final long days, final long months, final long years) {
        this.nanos = nanos;
        this.days = days;
        this.months = months;
        this.years = years;
    }

    public Duration() {
        this(0, 0, 0, 0);
    }

    public Duration(final Duration duration) {
        this(duration.nanos, duration.days, duration.months, duration.years);
    }

    public Duration(final java.time.Duration javaDuration) {
        this(javaDuration.getNano() + (javaDuration.getSeconds() % SECONDS_IN_DAY) * 1000000000L,
                javaDuration.getSeconds() / SECONDS_IN_DAY, 0, 0);
    }


    public Duration plusNanos(final long number) {
        return new Duration(this.nanos + number, this.days, this.months, this.years);
    }

    public Duration plusDays(final long number) {
        return new Duration(this.nanos, this.days + number, this.months, this.years);
    }

    public Duration plusMonths(final long number) {
        return new Duration(this.nanos, this.days, this.months + number, this.years);
    }

    public Duration plusYears(final long number) {
        return new Duration(this.nanos, this.days, this.months, this.years + number);
    }


    public Duration plus(final long number, final ChronoUnit chronoUnit) {
        switch (chronoUnit) {
            case NANOS:      return plusNanos(number);
            case MICROS:     return plusNanos(number * 1000L);
            case MILLIS:     return plusNanos(number * 1000000L);
            case SECONDS:    return plusNanos(number * 1000000000L);
            case MINUTES:    return plusNanos(number * 60000000000L);
            case HOURS:      return plusNanos(number * 3600000000000L);
            case HALF_DAYS:  return plusNanos(number * 43200000000000L); // 12 hours

            case DAYS:       return plusDays(number);
            case WEEKS:      return plusDays(number * 7L);

            case MONTHS:     return plusMonths(number);

            case YEARS:      return plusYears(number);
            case DECADES:    return plusYears(number * 10L);
            case CENTURIES:  return plusYears(number * 100L);
            case MILLENNIA:  return plusYears(number * 1000L);
            case ERAS:       return plusYears(number * 1000000000L);

            case FOREVER:
            default:
                throw new RuntimeException("Invalid chrono unit: " + chronoUnit);
        }
    }

    public Duration plus(final double number, final ChronoUnit chronoUnit) {
        final Duration result = plus(roundToLong(number), chronoUnit);
        switch (chronoUnit) {
            case NANOS:
            case MICROS:
            case MILLIS:
            case SECONDS:
            case MINUTES:
            case HOURS:
            case HALF_DAYS:
                return result;

            case DAYS:
            case WEEKS:
                // Convert to nanos directly, assuming 1 * day = DAY.getDuration().toNanos() * ns,
                // even though that's not always the case because of e.g. summer/winter time changes
                return result.plusNanos(
                        roundToLong(
                                remainderFromRoundingToLong(number)
                                        * chronoUnit.getDuration().toNanos()
                        )
                );

            case MONTHS:
                // Convert months to days using DAYS_IN_MONTH, which was obtained using the built-in
                // ChronoUnit estimations for month and day's average durations (in seconds)
                return result.plus(remainderFromRoundingToLong(number) * DAYS_IN_MONTH,
                        ChronoUnit.DAYS);

            case YEARS:
            case DECADES:
            case CENTURIES:
            case MILLENNIA:
            case ERAS:
                // According to the built-in ChronoUnit estimations, a year is exactly 12 months
                // (even though that's not always the case because e.g. of leap years), so
                // monthsInYearUnit is surely an integer
                final long monthsInYearUnit = chronoUnit.getDuration().getSeconds()
                        / ChronoUnit.MONTHS.getDuration().getSeconds();
                return result.plus(remainderFromRoundingToLong(number) * monthsInYearUnit,
                        ChronoUnit.MONTHS);

            case FOREVER:
            default:
                throw new RuntimeException("Invalid chrono unit: " + chronoUnit);
        }
    }

    public Duration plus(final Number number, final ChronoUnit chronoUnit) {
        return number.isInteger()
                ? plus(number.integerValue(), chronoUnit)
                : plus(number.decimalValue(), chronoUnit);
    }

    public Duration plus(final Duration duration) {
        return new Duration(
                nanos + duration.nanos,
                days + duration.days,
                months + duration.months,
                years + duration.years
        );
    }


    public Duration multiply(final long number) {
        return new Duration(nanos * number, days * number, months * number, years * number);
    }

    public Duration multiply(final double number) {
        return new Duration(roundToLong(nanos * number), 0, 0, 0)
                .plus(days * number, ChronoUnit.DAYS)
                .plus(months * number, ChronoUnit.MONTHS)
                .plus(years * number, ChronoUnit.YEARS);
    }

    public Duration multiply(final Number number) {
        return number.isInteger()
                ? multiply(number.integerValue())
                : multiply(number.decimalValue());
    }


    public long getNanos() {
        return nanos;
    }

    public long getDays() {
        return days;
    }

    public long getMonths() {
        return months;
    }

    public long getYears() {
        return years;
    }


    @Override
    public String toString() {
        return toJavaDuration().toString();
    }

    public java.time.Duration toJavaDuration() {
        return java.time.Duration.ofNanos(nanos)
                .plusDays(days)
                .plus(multiplyDurationBy(months, ChronoUnit.MONTHS))
                .plus(multiplyDurationBy(years, ChronoUnit.YEARS));
    }

    public LocalDateTime applyAsOffsetToDateTime(final LocalDateTime original) {
        return original.plusNanos(nanos)
                .plusDays(days)
                .plusMonths(months)
                .plusYears(years);
    }

    private java.time.Duration multiplyDurationBy(final long number,
                                                  final ChronoUnit chronoUnit) {
        // This operation would not be doable directly with Duration.plus(), since it throws an
        // exception with units with an estimated duration
        return chronoUnit.getDuration().multipliedBy(number);
    }
}

File: numbers/src/main/java/org/dicio/numbers/unit/Duration.kt
package org.dicio.numbers.unit

import org.dicio.numbers.util.Utils
import java.time.LocalDateTime
import java.time.temporal.ChronoUnit

/**
 * TODO add documentation and tests
 */
class Duration @JvmOverloads constructor(// basic well-defined unit of time, fits precisely up to ~10^3 hours
    @JvmField val nanos: Long = 0, // not stored as nanos because of e.g. summer/winter time changes
    @JvmField val days: Long = 0, // not stored as days because months have different numbers of days
    @JvmField val months: Long = 0, // not stored as months because of e.g. historical changes
    @JvmField val years: Long = 0
) {
    constructor(duration: Duration) : this(
        duration.nanos,
        duration.days,
        duration.months,
        duration.years
    )

    constructor(javaDuration: java.time.Duration) : this(
        javaDuration.nano + (javaDuration.seconds % SECONDS_IN_DAY) * 1000000000L,
        javaDuration.seconds / SECONDS_IN_DAY, 0, 0
    )


    fun plusNanos(number: Long): Duration {
        return Duration(this.nanos + number, this.days, this.months, this.years)
    }

    fun plusDays(number: Long): Duration {
        return Duration(this.nanos, this.days + number, this.months, this.years)
    }

    fun plusMonths(number: Long): Duration {
        return Duration(this.nanos, this.days, this.months + number, this.years)
    }

    fun plusYears(number: Long): Duration {
        return Duration(this.nanos, this.days, this.months, this.years + number)
    }


    fun plus(number: Long, chronoUnit: ChronoUnit): Duration {
        return when (chronoUnit) {
            ChronoUnit.NANOS -> plusNanos(number)
            ChronoUnit.MICROS -> plusNanos(number * 1000L)
            ChronoUnit.MILLIS -> plusNanos(number * 1000000L)
            ChronoUnit.SECONDS -> plusNanos(number * 1000000000L)
            ChronoUnit.MINUTES -> plusNanos(number * 60000000000L)
            ChronoUnit.HOURS -> plusNanos(number * 3600000000000L)
            ChronoUnit.HALF_DAYS -> plusNanos(number * 43200000000000L) // 12 hours

            ChronoUnit.DAYS -> plusDays(number)
            ChronoUnit.WEEKS -> plusDays(number * 7L)

            ChronoUnit.MONTHS -> plusMonths(number)

            ChronoUnit.YEARS -> plusYears(number)
            ChronoUnit.DECADES -> plusYears(number * 10L)
            ChronoUnit.CENTURIES -> plusYears(number * 100L)
            ChronoUnit.MILLENNIA -> plusYears(number * 1000L)
            ChronoUnit.ERAS -> plusYears(number * 1000000000L)

            ChronoUnit.FOREVER -> throw RuntimeException("Invalid chrono unit: $chronoUnit")
            else -> throw RuntimeException("Invalid chrono unit: $chronoUnit")
        }
    }

    fun plus(number: Double, chronoUnit: ChronoUnit): Duration {
        val result = plus(Utils.roundToLong(number), chronoUnit)
        when (chronoUnit) {
            ChronoUnit.NANOS, ChronoUnit.MICROS, ChronoUnit.MILLIS, ChronoUnit.SECONDS, ChronoUnit.MINUTES, ChronoUnit.HOURS, ChronoUnit.HALF_DAYS -> return result

            ChronoUnit.DAYS, ChronoUnit.WEEKS ->                 // Convert to nanos directly, assuming 1 * day = DAY.getDuration().toNanos() * ns,
                // even though that's not always the case because of e.g. summer/winter time changes
                return result.plusNanos(
                    Utils.roundToLong(
                        Utils.remainderFromRoundingToLong(number)
                                * chronoUnit.duration.toNanos()
                    )
                )

            ChronoUnit.MONTHS ->                 // Convert months to days using DAYS_IN_MONTH, which was obtained using the built-in
                // ChronoUnit estimations for month and day's average durations (in seconds)
                return result.plus(
                    Utils.remainderFromRoundingToLong(number) * DAYS_IN_MONTH,
                    ChronoUnit.DAYS
                )

            ChronoUnit.YEARS, ChronoUnit.DECADES, ChronoUnit.CENTURIES, ChronoUnit.MILLENNIA, ChronoUnit.ERAS -> {
                // According to the built-in ChronoUnit estimations, a year is exactly 12 months
                // (even though that's not always the case because e.g. of leap years), so
                // monthsInYearUnit is surely an integer
                val monthsInYearUnit = (chronoUnit.duration.seconds
                        / ChronoUnit.MONTHS.duration.seconds)
                return result.plus(
                    Utils.remainderFromRoundingToLong(number) * monthsInYearUnit,
                    ChronoUnit.MONTHS
                )
            }

            ChronoUnit.FOREVER -> throw RuntimeException("Invalid chrono unit: $chronoUnit")
            else -> throw RuntimeException("Invalid chrono unit: $chronoUnit")
        }
    }

    fun plus(number: Number, chronoUnit: ChronoUnit): Duration {
        return if (number.isInteger
        ) plus(number.integerValue(), chronoUnit)
        else plus(number.decimalValue(), chronoUnit)
    }

    fun plus(duration: Duration): Duration {
        return Duration(
            nanos + duration.nanos,
            days + duration.days,
            months + duration.months,
            years + duration.years
        )
    }


    fun multiply(number: Long): Duration {
        return Duration(nanos * number, days * number, months * number, years * number)
    }

    fun multiply(number: Double): Duration {
        return Duration(Utils.roundToLong(nanos * number), 0, 0, 0)
            .plus(days * number, ChronoUnit.DAYS)
            .plus(months * number, ChronoUnit.MONTHS)
            .plus(years * number, ChronoUnit.YEARS)
    }

    fun multiply(number: Number): Duration {
        return if (number.isInteger
        ) multiply(number.integerValue())
        else multiply(number.decimalValue())
    }


    override fun toString(): String {
        return toJavaDuration().toString()
    }

    fun toJavaDuration(): java.time.Duration {
        return java.time.Duration.ofNanos(nanos)
            .plusDays(days)
            .plus(multiplyDurationBy(months, ChronoUnit.MONTHS))
            .plus(multiplyDurationBy(years, ChronoUnit.YEARS))
    }

    fun applyAsOffsetToDateTime(original: LocalDateTime): LocalDateTime {
        return original.plusNanos(nanos)
            .plusDays(days)
            .plusMonths(months)
            .plusYears(years)
    }

    private fun multiplyDurationBy(
        number: Long,
        chronoUnit: ChronoUnit
    ): java.time.Duration {
        // This operation would not be doable directly with Duration.plus(), since it throws an
        // exception with units with an estimated duration
        return chronoUnit.duration.multipliedBy(number)
    }

    companion object {
        private val SECONDS_IN_DAY = ChronoUnit.DAYS.duration.seconds
        private val DAYS_IN_MONTH =
            ChronoUnit.MONTHS.duration.seconds.toDouble() / ChronoUnit.DAYS.duration.seconds.toDouble() // ~30.436875
    }
}
